src/catchup/CatchupManagerImpl.cpp

```diff
@@ -79,59 +89,65 @@ CatchupManagerImpl::processLedger(LedgerCloseData const& ledgerData)
     // First: if CatchupWork has started, just buffer and return early.
     if (mCatchupWork)
     {
+        // we don't want to buffer any ledgers at or below where we are catching
+        // up to
+        auto const& config = mCatchupWork->getCatchupConfiguration();
+        if (ledgerData.getLedgerSeq() <= config.toLedger())
+        {
+            return;
+        }
+
         addToSyncingLedgers(ledgerData);
         logAndUpdateCatchupStatus(true);
         return;
     }
 
-    // Next switch between 3 cases: starting buffering, waiting to
-    // buffer, and already buffering.
-    auto& hm = mApp.getHistoryManager();
-    uint32_t firstLedgerInBuffer;
-    if (mSyncingLedgers.empty())
-    {
-        if (hm.isFirstLedgerInCheckpoint(lastReceivedLedgerSeq))
-        {
-            // Case 1: At the ledger where we should start buffering.
-            addToSyncingLedgers(ledgerData);
-            firstLedgerInBuffer = lastReceivedLedgerSeq;
-        }
-        else
-        {
-            // Case 2: Still waiting to start buffering.
-            firstLedgerInBuffer =
-                hm.firstLedgerAfterCheckpointContaining(lastReceivedLedgerSeq);
-        }
-    }
-    else
-    {
-        // Case 3: Already buffering, keep buffering.
-        addToSyncingLedgers(ledgerData);
-        firstLedgerInBuffer = mSyncingLedgers.front().getLedgerSeq();
-    }
-    assert(hm.isFirstLedgerInCheckpoint(firstLedgerInBuffer));
+    // Next, we buffer every out of sync ledger to allow us to get back in sync
+    // in case the ledgers we're missing are received.
+    addToSyncingLedgers(ledgerData);
 
-    // Finally we wait some number of ledgers beyond the first buffered
+    // Finally we wait some number of ledgers beyond the buffered checkpoint
     // ledger before we trigger the CatchupWork. This could be any number,
     // for the sake of simplicity at the moment it's set to one ledger
-    // after the first buffered one.
+    // after the first buffered one. Since we can receive out of order ledgers,
+    // we make sure that we have the corresponding checkpoint for the largest
+    // buffered ledger
+
     std::string message;
-    uint32_t catchupTriggerLedger =
-        hm.ledgerToTriggerCatchup(firstLedgerInBuffer);
-    if (lastReceivedLedgerSeq >= catchupTriggerLedger)
+    auto& hm = mApp.getHistoryManager();
+
+    uint32_t lastLedgerInBuffer = mSyncingLedgers.crbegin()->first;
+    uint32_t firstLedgerInCheckpoint =
+        hm.firstLedgerInCheckpointContaining(lastLedgerInBuffer);
+
+    auto it = mSyncingLedgers.find(firstLedgerInCheckpoint);
+    if (it != mSyncingLedgers.end() &&
```

Comment:
@MonsieurNicolas I was thinking about ways to make this condition simpler, but I think looking for the start of the checkpoint based on the largest buffered ledger is the best way to do this. I could also address https://github.com/stellar/stellar-core/issues/2455 if you think it's worth doing now, which will change this code a bit.